pkg/noderesourcetopology/cache/store.go

```diff
@@ -0,0 +1,218 @@
+	for key, res := range rs {
+		// We cannot predict on which Zone the workload will be placed.
+		// And we should totally not guess. So the only safe (and conservative)
+		// choice is to decrement the available resources from *all* the zones.
+		// This can cause false negatives, but will never cause false positives,
+		// which are much worse.
+		for zi := 0; zi < len(nrt.Zones); zi++ {
+			zone := &nrt.Zones[zi] // shortcut
+			for ri := 0; ri < len(zone.Resources); ri++ {
+				zr := &zone.Resources[ri] // shortcut
+				qty, ok := res[corev1.ResourceName(zr.Name)]
+				if !ok {
+					// this is benign; it is totally possible some resources are not
+					// available on some zones (think PCI devices), hence we don't
+					// even report this error, being an expected condition
+					continue
+				}
+				if zr.Available.Cmp(qty) < 0 {
+					// this should happen rarely, and it is likely caused by
+					// a bug elsewhere.
+					klog.V(3).InfoS("nrtcache: cannot decrement resource", "logKey", logKey, "zone", zr.Name, "node", nrt.Name, "available", zr.Available, "requestor", key, "quantity", qty)
+					zr.Available = zeroQty
+					continue
+				}
+
+				zr.Available.Sub(qty)
+			}
+		}
+	}
+}
+
+type counter map[string]int
+
+func (cnt counter) Incr(key string) int {
+	val := cnt[key]
+	val++
+	cnt[key] = val
+	return val
+}
+
+func (cnt counter) IsSet(key string) bool {
+	return cnt[key] > 0
+}
+
+func (cnt counter) Delete(key string) {
+	delete(cnt, key)
+}
+
+// podFingerprintForNodeTopology extracts without recomputing the pods fingerprint from
+// the provided Node Resource Topology object.
+func podFingerprintForNodeTopology(nrt *topologyv1alpha1.NodeResourceTopology) string {
+	if nrt.Annotations == nil {
+		return ""
+	}
+	return nrt.Annotations[podfingerprint.Annotation]
+}
+
+// checkPodFingerprintForNode verifies if the given pods fingeprint (usually from NRT update) matches the
+// computed one using the stored data about pods running on nodes. Returns nil on success, or an error
+// describing the failure
+func checkPodFingerprintForNode(logKey string, indexer NodeIndexer, nodeName, pfpExpected string) error {
+	objs, err := indexer.PodNamespacedNamesByNode(logKey, nodeName)
+	if err != nil {
+		return err
+	}
+
+	klog.V(6).InfoS("nrtcache: podset fingerprint", "logKey", logKey, "pods", len(objs))
+
+	pfp := podfingerprint.NewFingerprint(len(objs))
+	for _, obj := range objs {
+		klog.V(6).InfoS("nrtcache: podset fingerprint", "logKey", logKey, "namespacedName", obj.Namespace+"/"+obj.Name)
+		pfp.Add(obj.Namespace, obj.Name)
+	}
+	pfpComputed := pfp.Sign()
+
+	klog.V(5).InfoS("nrtcache: podset fingerprint check", "logKey", logKey, "node", nodeName, "expected", pfpExpected, "computed", pfpComputed)
+
+	return pfp.Check(pfpExpected)
+}
+
+// resourceCountersFromPod computes the cumulative resource requests required by the given pod.
+func resourceCountersFromPod(pod *corev1.Pod) corev1.ResourceList {
```

Comment:
Maybe use https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/pkg/util/resource.go#L50 instead which covers init containers as well

Reply:
thanks this seems better indeed!